pkg/controller/kmeshsecurity/security_client.go

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2024 The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kmeshsecurity
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/types/known/structpb"
+	pb "istio.io/api/security/v1alpha1"
+	"istio.io/istio/pkg/security"
+	"istio.io/istio/pkg/spiffe"
+	nodeagentutil "istio.io/istio/security/pkg/nodeagent/util"
+	pkiutil "istio.io/istio/security/pkg/pki/util"
+	"kmesh.net/kmesh/pkg/controller/workload"
+	"kmesh.net/kmesh/pkg/nets"
+)
+ 
+ var tlsOpts *TLSOptions 
+
+ type CitadelClient struct {
+	 // It means enable tls connection to Citadel if this is not nil.
+	 tlsOpts  *TLSOptions
+	 client   pb.IstioCertificateServiceClient
+	 conn     *grpc.ClientConn
+	 opts     *security.Options
+ }
+
+ type TLSOptions struct {
+	RootCert string
+	Key      string
+	Cert     string
+}
+ 
+ // NewCitadelClient create a CA client for Citadel.
+ func NewCitadelClient(opts *security.Options, tlsOpts *TLSOptions) (*CitadelClient, error) {
+	 var err error;
+ 
+	 c := &CitadelClient{
+		 tlsOpts:  tlsOpts,
+		 opts:     opts,
+	 }
+
+	 conn, err := nets.GrpcConnect(CSRSignAddress);
+	 if err != nil {
+		 log.Errorf("Failed to connect to endpoint %s: %v", opts.CAEndpoint, err)
+		 return nil, fmt.Errorf("failed to connect to endpoint %s", opts.CAEndpoint)
+	 }
+ 
+	 c.conn = conn
+	 c.client = pb.NewIstioCertificateServiceClient(conn)
+	 return c, nil
+ }
+
+// CSRSign calls Citadel to sign a CSR.
+func (c *CitadelClient) CSRSign(csrPEM []byte, certValidTTLInSec int64) (res []string, err error) {
+	crMetaStruct := &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			security.CertSigner: {
+				Kind: &structpb.Value_StringValue{StringValue: c.opts.CertSigner},
+			},
+		},
+	}
+	req := &pb.IstioCertificateRequest{
+		Csr:              string(csrPEM),
+		ValidityDuration: certValidTTLInSec,
+		Metadata:         crMetaStruct,
+	}
+
+	defer func() {
+		if err != nil {
+			log.Errorf("failed to sign CSR: %v", err)
+			if err := c.reconnect(); err != nil {
```

Comment:
In the function each err is handled and return result. so will `reconnect` work here?
I don't think it's going to go into that if branch.